### DISCUSSION TITLE
Disable default libc

### DISCUSSION THREAD

**Discussion by robertapengelly** [open-watcom/open-watcom-v2#778] [General]
I'm trying to disable using the watcom libc.

Help shows:
-zl               remove default library information

But it's still looking into the watcom libc.  Is this a bug or am I using the wrong flag?

**Comment by robertapengelly** [comment]
Done a bit more testing and got the following:

C:\Users\robertapengelly\Desktop\test>wcl -ecc -q -w -c -mt -zl -fpi87 -s -zdp -zu temp.c

C:\Users\robertapengelly\Desktop\test>wlink F { temp.obj } Form dos com N test.bin
Open Watcom Linker Version 2.0 beta Nov 15 2021 06:30:30 (64-bit)
Copyright (c) 2002-2021 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
Error! E2028: printf_ is an undefined reference
Warning! W1023: no starting address found, using 0000:0000
creating a DOS .COM executable
file temp.obj(C:\Users\robertapengelly\Desktop\test\temp.c): undefined symbol printf_

The C file looks like:

#include	<stdio.h>

void lilo_main (void) {
    printf ("Hello, world!");
}

The behavour is somewhat what I expect where printf isn't found but I don't expect stdio.h to be found either unless I'm proving it myself, am I missing something?

GCC has the -nostdinc flag but I can't seem to find anything in wcl help other than -zl and -zld but they both find stdio.h.

**Comment by jmalak** [comment]
I don't know what you want to do.

If you want to use compiled code without OW CRTL then it can be a little harder then only remove reference to OW run-time libraries.
The reference to OW run-time libraries can be simply removed by -zl compiler option.
-zld option has nothing to do with libraries, it only doesn't emit auto dependency info into object files which is used by wmake.
You should use also -zls option to suppress emiting of some run-time libraries symbol reference.

**Comment by robertapengelly** [comment]
What I'm trying to do is not rely on the OW libc at all.  Lets say I'm creating an operating system and I'm creating my own libc I want to only get headers from my implementation not from OW.

**Comment by jmalak** [comment]
OK. 
In this case you will need probably to implement some code generator support routines or extract it from OW CRTL, it depend on your code if use floating point or 64-bit math and on architecture 16 or 32-bit etc.

**Comment by robertapengelly** [comment]
Yeah I get that but say I use typedef size_t long in stddef.h and use #include <string.h> without creating it then OW tries to load the string.h included with itself and complains that size_t is already defined instead of complaining that string.h doesn't exist.

**Comment by jmalak** [comment]
Sorry, but if you plan to have your own CRTL then you must create your own header files, because they are not independent.

**Comment by jmalak** [comment]
Maybe I didn't understand you. If you want to not use header files then cleanup INCLUDE environment variable which point to system header files directories or setup it to what you need.

**Comment by robertapengelly** [comment]
I know I need to create my own headers but the problem is that OW still looks in its lh folder for the headers as well which is what I'm trying to stop.  GCC has a -nostdinc option which means that it won't use any system headers and I'm asking if OW has a similar feature.  If I remove the environment variable then I won't be able to use OW in other projects unless I add it again which will be going round in circles remove it for one project, add it for another and the loop will continue.

**Comment by jmalak** [comment]
It is not possible, it signal some OW misconfiguration.
in OW doesn't exist any reference to lh subdirectory except compiler command line or INCLUDE environment variable which is part of your configuration.
you can use -x option to switch off usage of INCLUDE environment variables then only -i option can be used to specify include directories for header files.

**Comment by jmalak** [comment]
Anyway don't expect same solution as GCC.
WATCOM is much older then GCC that totally different approach and also OW is cross-compiler by nature.

**Comment by robertapengelly** [comment]
I weren't expecting the same solution I was expecting an OW equivalent.  I get that OW is older and a cross-compiler I just thought maybe there would be a flag to ignore the lh directory.

**Comment by jmalak** [comment]
it has nothing to do with lh directory.
compiler not reference it, only your configuration and ÿour compiler command line.
That you must remove it if you don't want it, compiler doesn't expect anything about header files.

**Comment by robertapengelly** [comment]
Trying to build on linux gives:

robertapengelly@RAP1505-UBUNTU:~/Desktop/test$ wcl -ecc -q -w -c -mt -zl -zdl -fpi87 -s -zdp -zu temp.c
/opt/watcom/lh/string.h(108): Error! E1034: Symbol 'size_t' already defined
/opt/watcom/lh/string.h(108): Note! N2002: 'size_t' defined in: temp.c(1)
Error: Compiler returned a bad status compiling 'temp.c'

The same happened earlier on Windows and thats installed using the installer on here.

**Comment by jmalak** [comment]
don't ask me to resolve your code bugs.
I saw what you must do to exclude system headers from compiler search path it doesn't mean that it will ignore direct reference in your code. 
I can say the system header files not reference any subdirectory that what you describe is your mistake not mistake in header files. If you include one header file it doesn't mean that no other header files is not included.
you have some mistake somewhere and I cannot find it for you.
If you exclude system headers from search path then you can not get such behavior, it signal that you do something wrong.

**Comment by robertapengelly** [comment]
I don't have code bugs I have like 3 lines of code:

typedef long size_t;
#include    <string.h>

void main (void) {}

I'm trying to get OW to ignore the default headers.

**Comment by jmalak** [comment]
this code is buggy, look into C language standard, and any compiler will report bug.
I say what you should do but you ignore it then it do what should do report bug.

**Comment by robertapengelly** [comment]
Take the code above I don't want it complain "Symbol 'size_t' already defined", I want a way that it will complain about string.h being missing instead given that I haven't provided one.

**Comment by jmalak** [comment]
you ignore what I wrote about INCLUDE environment variables. It is source of your "problem".

**Comment by robertapengelly** [comment]
No I haven't ignore it you ignored what I said earlier, lets say I remove the INCLUDE environment variable to compile a project using only my libc implementation then I want to make an application for dos using OW ohh wait I can't cause I removed the INCLUDE so I re-add it then I go back to the first project but hang on I got to remove the INCLUDE again it's going to be a never ending loop of adding and removing.  Surely there's an easier way than all that.

**Comment by groessler** [comment]
Either you have the INCLUDE env variable, or you don't have it. I don't see the problem. (I haven't tested and verified your problem.)

**Comment by jmalak** [comment]
below is result from Linux for your sample code if you apply what I recommended.

```
Open Watcom C x86 32-bit Optimizing Compiler
Version 2.0 beta Dec  2 2021 20:48:09 (32-bit)
Copyright (c) 2002-2021 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1984-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
tx1.c(2): Error! E1055: Unable to open 'string.h'
tx1.c: 6 lines, 0 warnings, 1 errors
```
I run it as I recommend you by command
`wcc386 -x tx1.c
`

**Comment by robertapengelly** [comment]
Ohh the x is for wcl I thought you were telling me to recompile the project as you said "It is not possible, it signal some OW misconfiguration".  Sorry a bit of misunderstanding.  I must have skipped past that flag when I was searching the help.  Problem solved.

**Comment by jmalak** [comment]
Here is appropriate part of OW documentation which describe include files processing in OW.
[Open Watcom C/C++  #include File Processing](https://open-watcom.github.io/open-watcom-v2-wikidocs/cguide.html#Open_Watcom_CDCPP__include_File_Processing)

**Comment by robertapengelly** [comment]
Passing -x to wcl did what I wanted so the problem is solved.  I must have missed it when I was looking at wcl help.

**Comment by jmalak** [comment]
It is difference OW 2 against OW1.9 that -x option works same way for all tools including resource compiler.
You can check if you run OW tool without command line arguments then show you available options.
By example WRC use same logic for include files and has -x option etc.

```
Open Watcom Windows and OS/2 Resource Compiler
Version 2.0 beta Nov 11 2021 19:19:52 (32-bit)
Copyright (c) 2002-2021 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1993-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
Usage: wrc {options} input-filename {options} [output-filename] {options}
Options:
-30              stamp file as requiring Windows 3.0
-31              (*) stamp file as requiring Windows 3.1
-?               print this message
-ad              generate auto dependency information for use by wmake
-bt=<os>         set the build target to <os> [windows|nt|os2]
-c=<file>        set code page conversion file
-d=<name>[=text] define text macro <name>[=text]
-e               for a DLL, global memory above EMS line
-fe=<file>       set the output executable file to name
-fo=<file>       set the output resource file to name
-i=<path>        look in path for include files
-k               don't sort segments (same as -s0)
-l               program uses LIM 3.2 EMS directly
-m               each instance of program has its own EMS bank
-o               preprocess only
-p               private DLL
-q               operate quietly
-r               only build the resource file
-s{0,1,2}        Segment and resource sorting method
  0              - no sorting, leave segments in the linker order
  1              - move preload segments to front and mark for fast load
  2              - (*) move preload, data, non-discard. segments to front
-t               protected mode only
-x               ignore the INCLUDE environment variable
        Multi-byte characters support
-zk{0,1,2,3,u0,u8} 
  0              - (*) Japanese (Kanji, CP 932)
  1              - Chinese (Traditional, CP 950)
  2              - Korean (Wansung, CP 949)
  3              - Chinese (Simplified, CP 936)
  u0             - Japanese (Kanji, CP 932), UTF-8 source
  u8             - Unicode UTF-8
-zm              output Microsoft/IBM format .res files
-zn              don't preprocess the file
```

**Comment by jmalak** [comment]
WCL tool is not suitable for your purpose now, because it use predefined linker definition for OW run-time libraries.
If you want to use wcl and OW linker then you will need to create your own wlink.lnk  definition file to link executable with your libraries.

**Comment by robertapengelly** [comment]
Yeah I know about wlink e.g. wlink F { temp.obj } Form dos com N test.sys.  I just didn't notice the -x flag in wcl nor wcc help.

**Comment by jmalak** [comment]
No, -x is useful in both.
But wcc is C compile only tool and wcl is "compile and link".
wcl use wlink but works with standard wlink definitions in wlink.lnk but you need control over full wlink command that without changes in wlink.lnk it can do something what you not expect.
Simplest way is add new link target to wlink.lnk with your specific linker setup and next use it with wcl and wlink tools. 
Take into account that full link control from wcl is possible but it's unnecessarily complicated.
definion of new target is much simpler.

By example, we use for OW build custom build/wlink.lnk which is completely different from standard distribution.
There are special targets, one of for build loadable module for Linux but with REX format it uses special dynamic loader for such modules.
If it is linked then it is linked by standard tools as wlink with directive 
`system rex_linux`

or by wcl with option -l=rex_linux

By default such target is not supported by OW.

Appropriate definition of such target in wlink.lnk is
```
system begin rex_linux
    option osname='Linux 32-bit Pharlap REX'
    libpath %OWSRCDIR%/clib/library/linux.386/mf_r
    libpath %OWSRCDIR%/mathlib/library/msdos.386/mf_r
    disable 1014,1023
    format pharlap rex ^
end
```

**Comment by robertapengelly** [comment]
Not related to my problem the other day but considering this is still open how do you define variables for WASM.

I've tried both "%define temp 4" and "#define temp 4" but they both give me "test.asm(9): Error! E032: Syntax error".

**Comment by jmalak** [comment]
Such constructs are invalid.
First one is probably from some other assembler and second one is from C language.
WASM use MASM 6 assembler syntax.
you need to learn MASM assembler to understand how to define variables in WASM assembler.

**Comment by robertapengelly** [comment]
The first one is from NASM and I know the second one was from C but thought I'd try it anyway.  I know that MASM has = and equ which work great for numerical values but say I wanted test = (bp - 2) then that's not numerical and I get Error! E069: Illegal use of register for both = and equ.  If I use equ in NASM i get "error: bad syntax for EQU" which is where the %define test (bp - 2) comes in handy and I'm asking if WASM (or MASM) has an alternative or would I be better off sticking with NASM?

**Comment by jmalak** [comment]
It is different world, you must decided what you want.
Anyway I think you have proper definition but what you get is message about wrong instruction where you put this macro.
You need to understand MASM syntax of instruction (it uses Intel form).
Probably you use bp register or bp register expression where it is not permited etc.

**Comment by robertapengelly** [comment]
I'm not using it anywhere take the following:

% .model memodel

temp = (bp - 2)

_TEXT   segment     public 'CODE'
public _cstart_
_cstart_:
_TEXT   ends

end     _cstart_

I'm getting 'test.wasm(3): Error! E069: Illegal use of register' which indicates that I can't use bp with = and I've searched the web for over the last couple of days but can't find anything that helps.

**Comment by jmalak** [comment]
This 3 line is source of problem, it is garbage in MASM 
% .model memodel

temp = (bp - 2)

You need to learn MASM macro assembler, 
It has nothing to do with NASM syntax.
These lines is nonsense in MASM that it is hard to say what message you get.

**Comment by robertapengelly** [comment]
I'm trying to use WASM not MASM.

% .model memodel

temp = 2

_TEXT   segment     public 'CODE'
public _cstart_
_cstart_:
_TEXT   ends

end     _cstart_

Compiles no problem with 'wasm -zq -zcm -fo=.obj -Dmemodel=tiny test.wasm'

% .model memodel

temp = (bp - 2)

_TEXT   segment     public 'CODE'
public _cstart_
_cstart_:
_TEXT   ends

end     _cstart_

Gives me 'test.wasm(3): Error! E069: Illegal use of register'.

I'm asking if there's an alternative to NASM's %define so I can use a register and offset (e.g. bp - 2) instead of hard coding it where they are needed?

**Comment by robertapengelly** [comment]
I've finally solved the problem partly, if WASM is MASM syntax then how come the following doesn't build?

;% .model memodel

TxtEqu  equ <11[bp]>

_TEXT   segment     word    public 'CODE'
org     0
top:

public start
start proc

    jmp     short   bypass
    nop

bpb db 03ch dup(?)

bypass:

    mov     ax,     07c0h
    mov     ds,     ax
    
    xor     ax,     ax
    mov     ss,     ax
    
    mov     bp,     7c00h
    lea     sp,     [bp]
    
    mov     ax,     TxtEqu
    call    writehex
    
    xor     ax,     ax
    int     16h
    int     19h

start endp

writehex proc

    push    ax
    push    bx
    push    cx
    push    di
    
    mov     cl,     4
    mov     di,     4

writehex_loop:

    rol     ax,     cl
    push    ax
    and     al,     00001111b
    cmp     al,     10
    jae     writehex_high

writehex_low:

    add     al,     30h
    jmp     short   writehex_ischar

writehex_high:

    add     al,     37h

writehex_ischar:

    mov     ah,     0Eh
    mov     bx,     0007h
    int     10h
    
    pop     ax
    
    dec     di
    jnz     writehex_loop

writehex_done:

    pop     di
    pop     cx
    pop     bx
    pop     ax
    ret

writehex endp

org     01FEh
dw      0aa55h

_TEXT ends

end top

MASM compiled and does as expected but wasm gives me:

robertapengelly@RAP1505-UBUNTU:~/Desktop/test 01/temp$ wasm -zq -zcm -fo=.obj -Dmemodel=tiny TEST.ASM
TEST.ASM(28): Error! E039: Operands must be the same size
TEST.ASM(28): Warning! W041: Immediate constant too large
